File: packages/python/src/process/process_registry.py
from dataclasses import dataclass
import subprocess

from src.dataclass.process import ProcessInfo


@dataclass
class ProcessRecord:
    """Dataclass to define a singular process record"""

    info: ProcessInfo
    instance: subprocess.Popen[str] | subprocess.Popen[bytes]


class ProcessRegistry:
    """A ProcessRegistry is a class with a dictionary whose key is the pid,
    and the value is the metadata and its process instance"""

    def __init__(self) -> None:
        self._registry: dict[int, ProcessRecord] = {}

    def add(
        self,
        info: ProcessInfo,
        instance: subprocess.Popen[str] | subprocess.Popen[bytes],
    ) -> None:
        """Add a record into the registry, ensuring the record has not already been added"""
        if info.pid in self._registry:
            raise ValueError(
                f"Registry entry with pid {info.pid} already exists in registry: {self._registry[info.pid].info.command}"
            )
        self._registry[info.pid] = ProcessRecord(info=info, instance=instance)

    def remove(self, pid: int) -> bool:
        """Remove a record from the registry, ensuring the record exists in the registry"""
        if pid not in self._registry:
            return False
        del self._registry[pid]
        return True

    def get(self, pid: int) -> ProcessRecord | None:
        """Return a record from the registry"""
        return self._registry[pid]

    def list(self) -> list[ProcessRecord]:
        """Return record values as a list"""
        return list(self._registry.values())

    def exits(self, pid: int) -> bool:
        """Check if a process exists in the registry"""
        return pid in self._registry

    def clear(self) -> int:
        """Clears the registry and returns the number of records cleared"""
        count = len(self._registry)
        self._registry.clear()
        return count

    def count(self) -> int:
        """Returns the count of processes in the registry"""
        return len(self._registry)

File: packages/node/source/classes/process/ProcessRegistry.ts
import type { ChildProcess } from "node:child_process";

import type { IProcessInfo } from "@/types/process";

/**
 * A class that provides methods for managing records of active processes.
 *
 * @template P - The specific type of child process being managed.
 * @documentation [view on GitHub](https://github.com/octovel/environment/blob/stable/docs/guides/process-registry.md)
 */
export class ProcessRegistry<P extends ChildProcess = ChildProcess> {
  /**
   * Maps process IDs (`pid`) to their corresponding metadata and instance.
   *
   * @internal This property is used internally to store and manage process information.
   */
  private readonly registry = new Map<number, IProcessInfo & { instance: P }>();

  //#region Core Operations
  /**
   * Registers a process and its associated metadata into the private `registry` property.
   * Cannot add duplicate entries with the same Process ID.
   *
   * @param info - Metadata describing the process (name, args, timestamps, etc.).
   * @param instance - The Node.js {@link ChildProcess} instance.
   */
  public add(info: IProcessInfo, instance: P): void {
    if (this.registry.has(info.pid)) {
      throw new Error(`Process with PID ${info.pid} is already registered.`);
    }
    this.registry.set(info.pid, { ...info, instance });
  }

  /**
   * Removes a process from the registry by its PID.
   *
   * @param pid - The ID of the process to remove.
   * @returns `true` if the process was removed successfully, `false` otherwise.
   */
  public remove(pid: number): boolean {
    return this.registry.delete(pid);
  }

  /**
   * Retrieves process details by PID.
   *
   * @param pid - The process ID to retrieve.
   * @returns The associated process info and instance, or `undefined` if not found.
   */
  public get(
    pid: number,
  ): Readonly<IProcessInfo & { instance: P }> | undefined {
    const record = this.registry.get(pid);
    return record ? Object.freeze(record) : undefined;
  }

  /**
   * Returns all registered processes as an array of readonly objects.
   *
   * @returns A readonly array of process records.
   */
  public list(): readonly Readonly<IProcessInfo & { instance: P }>[] {
    return Array.from(this.registry.values()) as Readonly<
      IProcessInfo & { instance: P }
    >[];
  }

  /**
   * Checks whether a process with the given PID exists in the registry.
   *
   * @param pid - The process ID to check.
   * @returns `true` if the process exists, `false` otherwise.
   */
  public exists(pid: number): boolean {
    return this.registry.has(pid);
  }
  //#endregion

  //#region Utilities
  /**
   * Clears all entries in the process registry.
   *
   * @returns The number of processes removed.
   */
  public clear(): number {
    const count = this.registry.size;
    this.registry.clear();
    return count;
  }

  /**
   * Returns the total number of processes currently registered.
   *
   * @returns The total number of processes registered.
   */
  public count(): number {
    return this.registry.size;
  }
  //#endregion
}
